sway-ir/src/optimize/dce.rs

```diff
@@ -0,0 +1,158 @@
+                name: _,
+                params,
+                coins,
+                asset_id,
+                gas,
+            } => vec![*params, *coins, *asset_id, *gas],
+            Instruction::ExtractElement {
+                array,
+                ty: _,
+                index_val,
+            } => vec![*array, *index_val],
+            Instruction::ExtractValue {
+                aggregate,
+                ty: _,
+                indices: _,
+            } => vec![*aggregate],
+            Instruction::GetStorageKey => vec![],
+            Instruction::Gtf {
+                index,
+                tx_field_id: _,
+            } => vec![*index],
+            Instruction::GetPointer {
+                base_ptr: _,
+                ptr_ty: _,
+                offset: _,
+            } =>
+            // TODO: Not sure.
+            {
+                vec![]
+            }
+            Instruction::InsertElement {
+                array,
+                ty: _,
+                value,
+                index_val,
+            } => vec![*array, *value, *index_val],
+            Instruction::InsertValue {
+                aggregate,
+                ty: _,
+                value,
+                indices: _,
+            } => vec![*aggregate, *value],
+            Instruction::IntToPtr(v, _) => vec![*v],
+            Instruction::Load(v) => vec![*v],
+            Instruction::Nop => vec![],
+            Instruction::Phi(ins) => ins.iter().map(|v| v.1).collect(),
+            Instruction::ReadRegister(_) => vec![],
+            Instruction::Ret(v, _) => vec![*v],
+            Instruction::StateLoadQuadWord { load_val, key } => vec![*load_val, *key],
+            Instruction::StateLoadWord(key) => vec![*key],
+            Instruction::StateStoreQuadWord { stored_val, key } => vec![*stored_val, *key],
+            Instruction::StateStoreWord { stored_val, key } => vec![*stored_val, *key],
+            Instruction::Store {
+                dst_val,
+                stored_val,
+            } => {
+                vec![*dst_val, *stored_val]
+            }
+        }
+    }
+
+    // Go through each instruction and update use_count.
+    for (block, inst) in function.instruction_iter(context) {
+        let opds = get_operands(inst.get_instruction(context).unwrap());
+        for v in opds {
+            match context.values[v.0].value {
+                ValueDatum::Instruction(_) => {
+                    num_uses
+                        .entry(v)
+                        .and_modify(|(_block, count)| *count += 1)
+                        .or_insert((block, 1));
+                }
+                ValueDatum::Constant(_) | ValueDatum::Argument(_) => (),
+            }
+        }
+    }
+
+    let mut worklist: Vec<(Value, Block)> = Vec::new();
+    for (block, inst) in function.instruction_iter(context) {
+        if num_uses.get(&inst).is_none() {
+            worklist.push((inst, block));
+        }
+    }
```

Comment:
`worklist` could be constructed with a `filter` and `collect` and `num_uses` above could be constructed using `fold`.  I generally prefer the 'functional' methods for creating lists, rather an imperative update-a-mutable method, but that could just be me.

I feel that avoiding the creation of a mutable value is safer code and could probably be better optimised, but yeah, it's subjective.

Reply:
This is simple and straightforward enough to not bring in safety ambiguities and also quite readable.

Reply:
I'd argue that this is more readable too.
```rust
let worklist = function.instruction_iter(context)
    .filter(|(_block, inst)| num_used.get(&inst).is_none())
    .collect::<Vec<_>>()
```

Reply:
In any case, the mutable variable itself can't be avoided here as it's a worklist that changes as the algorithm progresses.

Reply:
> I'd argue that this is more readable too.
> 
> ```rust
> let worklist = function.instruction_iter(context)
>     .filter(|(_block, inst)| num_used.get(&inst).is_none())
>     .collect::<Vec<_>>()
> ```

Done.